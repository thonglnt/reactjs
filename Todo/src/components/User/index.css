.user-item {
  padding: 15px;
  border: 2px solid var(--charleston-green);
  display: flex;
  align-items: center;
  color: var(--white);
  cursor: pointer;
}

.user-selected {
  background-color: var(--ruddy-pink);
}

.user-item:hover {
  background-color: var(--ruddy-pink);
}

.user-item:hover .icon {
  display: block;
}

.icon-delete {
  display: none;
}

@media screen and (max-width: 992px) {
  .user-item {
    padding: 10px;
  }
}

@media screen and (max-width: 992px) {
  .user-item {
    padding: 5px;
  }

  .icon-delete {
    max-width: 10px;
    padding-left: 2px;
  }
}
