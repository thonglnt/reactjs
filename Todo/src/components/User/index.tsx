// Lib
import { NavigateFunction, useNavigate } from 'react-router-dom';

// Assets
import { iconDelete } from '@assets';

// Components
import Spinner from '@components/Spinner';

export type UserProps = {
  id: string;
  avatar: string;
  name: string;
  className: string;
  onRemoveUser: (id: string, name: string) => void;
  isRemovingUser?: boolean;
};

const User = ({ id, name, avatar, className, onRemoveUser, isRemovingUser = false }: UserProps) => {
  const navigate: NavigateFunction = useNavigate();
  
  const handleRemoveUser = () => {
    onRemoveUser(id, name);
  };

  const navigateUserSelected = () => {
    navigate(`/${id}`);
  };

  return (
    <li className={className} onClick={navigateUserSelected} data-testid='selected-id-user'>
      <img src={avatar} alt='Avatar' className='avatar' />
      {name.length > 10 ? `${name.substring(0, 10)}...` : name}
      {isRemovingUser ? (
        <Spinner className='icon' />
      ) : (
        <img
          src={iconDelete}
          onClick={handleRemoveUser}
          alt='user name'
          className='icon icon-delete'
          data-testid='delete-user'
        />
      )}
    </li>
  );
};

export default User;
