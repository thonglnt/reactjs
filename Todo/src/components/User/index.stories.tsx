// Lib
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';

// Components
import User, { UserProps } from '@components/User';
import { BrowserRouter } from 'react-router-dom';

export default {
  title: 'User',
  component: User,
};

const Template: Story<UserProps> = (args) => (
  <BrowserRouter>
    <User {...args} />
  </BrowserRouter>
);

export const UserSample = Template.bind({});
UserSample.args = {
  id: '1',
  avatar:
    'https://www.cgv.vn/media/catalog/product/cache/1/image/1800x/71252117777b696995f01934522c402d/a/v/avatar-1615695904-2089-1615696022.jpg',
  name: 'name',
  className: 'user-item',
  onRemoveUser: action('onRemoveUser'),
};

export const UserSelected = Template.bind({});
UserSelected.args = {
  id: '1',
  avatar:
    'https://www.cgv.vn/media/catalog/product/cache/1/image/1800x/71252117777b696995f01934522c402d/a/v/avatar-1615695904-2089-1615696022.jpg',
  name: 'name',
  className: 'user-item user-selected',
  onRemoveUser: action('onRemoveUser'),
};
