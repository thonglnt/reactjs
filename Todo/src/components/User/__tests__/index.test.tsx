// Lib
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import User, { UserProps } from '@components/User';

describe('Test User component', () => {
  const initUser: UserProps = {
    id: '1',
    avatar: 'avatar',
    name: 'thong',
    className: 'className',
    onRemoveUser: jest.fn(),
  };

  test('It should render correctly with initial Props', () => {
    const { container } = render(
      <BrowserRouter>
        <User {...initUser} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  test('It should match DOM Snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <User {...initUser} isRemovingUser={true} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  test('It should call remove user', () => {
    render(
      <BrowserRouter>
        <User {...initUser} />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByTestId('delete-user'));

    expect(initUser.onRemoveUser).toHaveBeenCalledTimes(1);
  });
});
