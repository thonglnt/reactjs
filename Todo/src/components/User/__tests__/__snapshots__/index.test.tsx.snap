// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test User component It should match DOM Snapshot 1`] = `
<div>
  <li
    class="className"
    data-testid="selected-id-user"
  >
    <img
      alt="Avatar"
      class="avatar"
      src="avatar"
    />
    thong
    <div
      class="loader icon"
    />
  </li>
</div>
`;

exports[`Test User component It should render correctly with initial Props 1`] = `
<div>
  <li
    class="className"
    data-testid="selected-id-user"
  >
    <img
      alt="Avatar"
      class="avatar"
      src="avatar"
    />
    thong
    <img
      alt="user name"
      class="icon icon-delete"
      data-testid="delete-user"
      src="IMAGE_MOCK"
    />
  </li>
</div>
`;
